src/net/send_recv.rs

```diff
@@ -133,4 +144,585 @@ pub fn sendto_unix<Fd: AsFd>(
+pub struct RecvMsgV6 {
+    /// The socket addr, only set if the socket was not bound before.
+    pub addr: Option<SocketAddrV6>,
+    /// How many bytes have been received.
+    pub bytes: usize,
+    /// The returned flags.
+    pub flags: RecvFlags,
+}
+
+impl RecvMsgV6 {
+    /// Safety: `msg` must be a valid return value from an Ipv6 based `recvmsg` call.
+    #[cfg(not(windows))]
+    pub(crate) unsafe fn new(
+        bytes: usize,
+        msg: imp::c::msghdr,
+        ancillary: Option<&mut RecvSocketAncillaryV6<'_>>,
+    ) -> Self {
+        let addr = read_sockaddr_v6_opt(msg.msg_name as *const _, msg.msg_namelen as _);
+        let flags = RecvFlags::from_bits_truncate(msg.msg_flags);
+
+        if let Some(ancillary) = ancillary {
+            ancillary.length = msg.msg_controllen as usize;
+            ancillary.truncated = flags.contains(RecvFlags::CTRUNC);
+        }
+
+        RecvMsgV6 {
+            bytes: bytes as usize,
+            addr,
+            flags,
+        }
+    }
+
+    /// Safety: `msg` must be a valid return value from an Ipv6 based `recvmsg` call.
+    #[cfg(windows)]
+    pub(crate) unsafe fn new(bytes: usize, msg: imp::c::msghdr) -> Self {
+        let addr = read_sockaddr_v6_opt(msg.msg_name as *const _, msg.msg_namelen as _);
+        let flags = RecvFlags::from_bits_truncate(msg.msg_flags);
+
+        RecvMsgV6 {
+            bytes: bytes as usize,
+            addr,
+            flags,
+        }
+    }
+}
+
+/// `recvmsg(fd, iovs, flags)`—Reads data from a socket.
+///
+/// # References
+///  - [POSIX]
+///  - [Linux]
+///
+/// [POSIX]: https://pubs.opengroup.org/onlinepubs/9699919799/functions/recvmsg.html
+/// [Linux]: https://man7.org/linux/man-pages/man2/recvmsg.2.html
+#[doc(alias = "recvmsg")]
+#[cfg(not(windows))]
+#[inline]
+pub fn recvmsg_unix<Fd: AsFd>(
+    fd: &Fd,
+    iovs: &[io::IoSliceMut<'_>],
+    flags: RecvFlags,
+) -> io::Result<RecvMsgUnix> {
+    let fd = fd.as_fd();
+    imp::syscalls::recvmsg_unix(fd, iovs, None, flags)
+}
+
+/// `recv(fd, iovs, flags)`—Reads data from a socket.
+///
+/// # References
+///  - [POSIX]
+///  - [Linux]
+///
+/// [POSIX]: https://pubs.opengroup.org/onlinepubs/9699919799/functions/recvmsg.html
+/// [Linux]: https://man7.org/linux/man-pages/man2/recvmsg.2.html
+#[doc(alias = "recvmsg")]
+#[cfg(not(windows))]
+#[inline]
+pub fn recvmsg_unix_with_ancillary<Fd: AsFd>(
+    fd: &Fd,
+    iovs: &[io::IoSliceMut<'_>],
+    ancillary: &mut RecvSocketAncillaryUnix<'_>,
+    flags: RecvFlags,
+) -> io::Result<RecvMsgUnix> {
+    let fd = fd.as_fd();
+    imp::syscalls::recvmsg_unix(fd, iovs, Some(ancillary), flags)
+}
+
+/// Return value from calling `recvmsg`.
+#[cfg(not(windows))]
+pub struct RecvMsgUnix {
+    /// The socket addr, only set if the socket was not bound before.
+    pub addr: Option<SocketAddrUnix>,
+    /// How many bytes have been received.
+    pub bytes: usize,
+    /// The returned flags.
+    pub flags: RecvFlags,
+}
```

Comment:
I modeled it after `sendto_*` which asfaict does not have a unified version either. I don't have a strong preference in adding it or not adding it either way, but it would be good to be consistent in the API overall.

Reply:
Sending is asymmetric from receiving here because sending always knows the address family of what it's sending up front. Receiving can't always know what it's receiving until after it's in the call. So we need a `SocketAddrAny` version of `recvmsg`, just as we need the `SocketAddrAny` version of `recvfrom`.

We may also want to add `SocketAddrAny` versions of `sendto` and `sendmsg`, but strictly speaking, that would be for convenience. (Or if there's a different way to handle `SocketAddrAny` overall, I'm open to that too.)

Reply:
right, that makes sense, I‘ll look into it

Reply:
added a `recvmsg` variant